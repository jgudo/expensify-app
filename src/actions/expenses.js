import uuid from 'uuid';
import database from '../firebase/firebase';

//component calls action generator 
// action generator return object
//component dispatches object
//redux store changes

//component calls action generator 
// action generator return function
//component dispatches function (?)
//function runs (has the ability to dispatch other actions and do whatever it wants)

//ADD_EXPENSE
export const addExpense = (expenses) => ({
    type: 'ADD_EXPENSE',
    expenses
});

//BEFORE -- without firebase
// export const addExpense = ( {description = '', note = '', amount = 0, createdAt = 0} = {} ) => ({
//     type: 'ADD_EXPENSE',
//     expenses: {
//         id: uuid(),
//         description,
//         note,
//         amount,
//         createdAt
//     }
// });

export const startAddExpense = (expenseData = {}) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        const {description = '', note = '', amount = 0, createdAt = 0} = expenseData;
        const expenses = {
            description,
            note,
            amount,
            createdAt
        };
    return database.ref(`users/${uid}/expenses`).push(expenses).then( (ref) => {
            dispatch(addExpense({
                id: ref.key,
                ...expenses
            }));
        });

    }
};


//EDIT_EXPENSE
export const editExpense = (id, updates) => ({
    type: 'EDIT_EXPENSE',
    id,
    updates
});

export const startEditExpense = (id, updates) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        return database.ref(`users/${uid}/expenses/${id}`).update(updates).then(() => {
            dispatch(editExpense(id, updates));
        });
    };
};

//REMOVE_EXPENSE

export const removeExpense = ( {id} ) => ({
    type: 'REMOVE_EXPENSE',
    id
});

export const startRemoveExpense = ({id}) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        return database.ref(`users/${uid}/expenses/${id}`).remove().then(() => {
            dispatch(removeExpense({ id }));
        });   
    };
};

//SET EXPENSES
export const setExpense = (expenses) => ({
    type: 'SET_EXPENSE',
    expenses
});

export const startSetExpense = () => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        return database.ref(`users/${uid}/expenses`).once('value').then((snapshot) => {
            const expenses = [];
            snapshot.forEach((childSnapshot) => {
                expenses.push({
                    id: childSnapshot.key,
                    ...childSnapshot.val()
                });
            });
            //console.log(expensesArr);
            dispatch(setExpense(expenses));
        });
    }
}
