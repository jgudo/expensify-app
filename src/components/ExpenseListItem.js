import React from 'react';
import {Link} from 'react-router-dom';
import moment from 'moment';
import numeral from 'numeral';

const ExpenseListItem = ({id,description, amount, createdAt}) =>  (
    <Link to={`/edit/${id}`} className="list-item">
        <div className="list-item__description">
            <h3 className="list-item__title">{description}</h3>
            <span>{moment(createdAt).format('MMMM Do YYYY')}</span>
        </div>
        <div>
            <h3> ₱{numeral(amount / 100).format('₱0,0.00')}</h3>
        </div>
    </Link>
);

export default ExpenseListItem;
//export default ExpenseListItem;