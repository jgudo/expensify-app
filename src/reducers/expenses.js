//Reducer 
const expensesReducer = (state = [], action) => {
    switch(action.type) {
        case 'ADD_EXPENSE': 
            return [...state, action.expenses];
        
        case 'REMOVE_EXPENSE': 
            return state.filter( (exp) => exp.id !== action.id );
        
        case 'EDIT_EXPENSE': 
            return state.map((expense) => {
                if(expense.id === action.id) {
                    return {
                        ...expense,
                        ...action.updates
                    };
                } else {
                    return expense;
                };
            });
        default: 
            return state;
    }
}

export default expensesReducer;