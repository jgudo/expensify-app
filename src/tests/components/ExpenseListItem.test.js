import React from 'react';
import {shallow} from 'enzyme';
import ExpenseListItem from '../../components/ExpenseListItem';
import expenses from '../fixtures/expenses';

test('Should render Expense list item', () => {
    const wrapper = shallow(<ExpenseListItem {...expenses[2]} />);
    expect(wrapper).toMatchSnapshot();
});