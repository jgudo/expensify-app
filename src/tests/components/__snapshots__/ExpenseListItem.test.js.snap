// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render Expense list item 1`] = `
<div>
  <Link
    replace={false}
    to="/edit/3"
  >
    <h3>
      Water Bill
    </h3>
  </Link>
  <p>
    $12.41
     - 
    January 4th 1970
  </p>
</div>
`;
